pkg/restore/restore_crd.go

```diff
@@ -0,0 +1,237 @@
+/*
+Copyright 2023 The Rook Authors. All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package restore
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"github.com/rook/kubectl-rook-ceph/pkg/debug"
+	"github.com/rook/kubectl-rook-ceph/pkg/exec"
+	"github.com/rook/kubectl-rook-ceph/pkg/k8sutil"
+	"github.com/rook/kubectl-rook-ceph/pkg/logging"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+var getCrName = `
+	kubectl -n %s get %s -o json | jq -r '.items[] | select(.metadata.deletionGracePeriodSeconds == 0) | .metadata.name'
+`
+
+var getCrYamlContent = `
+	kubectl -n %s get %s %s -oyaml > %s.yaml
+`
+
+var getK8sYamlContent = `
+	kubectl -n %s get %s -oyaml > %s
+`
+
+var getCrUID = `
+	kubectl -n %s get %s %s -o 'jsonpath={.metadata.uid}'
+`
+
+var removeFinalizers = `
+	kubectl -n %s patch %s/%s --type json --patch='[ { "op": "remove", "path": "/metadata/finalizers" } ]'
+`
+
+var recreateCr = `
+	kubectl create -f %s
+`
+
+func RestoreCrd(ctx context.Context, k8sclientset *k8sutil.Clientsets, operatorNamespace, clusterNamespace string, args []string) {
+	cr := args[0]
+	command := fmt.Sprintf(getCrName, clusterNamespace, cr)
+	crName := strings.TrimSpace(exec.ExecuteBashCommand(command))
+
+	if len(args) == 2 {
+		crName = args[1]
+	}
+
+	logging.Info("Restoring crd %s name %s", cr, crName)
+
+	logging.Info("Scaling down the operator to 0")
+	err := debug.SetDeploymentScale(ctx, k8sclientset.Kube, operatorNamespace, "rook-ceph-operator", 0)
+	if err != nil {
+		logging.Fatal(err)
+	}
+
+	logging.Info("Backing up kubernetes and crd resources")
+	command = fmt.Sprintf(getCrYamlContent, clusterNamespace, cr, crName, cr)
+	exec.ExecuteBashCommand(command)
+	logging.Info("Backed up crd %s named %s in file %s.yaml", cr, crName, cr)
+
+	command = fmt.Sprintf(getK8sYamlContent, clusterNamespace, "secret", "secret.yaml")
+	exec.ExecuteBashCommand(command)
+	logging.Info("Backed up secret in file secret.yaml")
+
+	command = fmt.Sprintf(getK8sYamlContent, clusterNamespace, "configmap", "configmap.yaml")
+	exec.ExecuteBashCommand(command)
+	logging.Info("Backed up configmap in file configmap.yaml")
```

Comment:
If we have full confidence on the automation to restore the deleted CRs, I don't think we need to backup the configmaps or secrets. The important configmaps and secrets anyway should have their own finalizers to prevent them from being deleted. The important thing will be to remove the ownerReferences from all resources so they won't be deleted when the parent resource is deleted.

Reply:
I have confidence in the too to restore based on my testing yet and also it's better to remove the secret/configmap backup as you mentioned these already have the finalizers